<launch>
	<arg name="postfix" default="" />
	<arg name="fixed_frame" default="base_link" />
	<param name="use_sim_time" value="true" />
	
	<node name="scan_to_cloud2" pkg="sq_lidar" type="sq_lidar">
		<param name="fixed_frame" value="$(arg fixed_frame)" />
		<remap from="~/scan" to="/sq_lidar$(arg postfix)/scan" />
		<remap from="~/cloud" to="/cloud" />
		<!-- <remap from="~/cloud" to="/cloud" /> -->
	</node>
	
	<include file="$(find beginner_tutorials)/launch/lcl.launch"/>

	<!-- <node pkg="beginner_tutorials" name="imu_alignment" type="imu_alignment" output="screen"/> -->
	<!-- <node pkg="beginner_tutorials" name="imu_alignment_simple" type="imu_alignment_simple" output="screen"/> -->
	<node pkg="beginner_tutorials" name="imu_alignment_bias" type="imu_alignment_bias" output="screen"/>

	<group ns="imu_slam_walls">
		<arg name="posemsg_name" default="/pose_imu_slam_walls"/>

		<node pkg="beginner_tutorials" name="sq2_normals" type="sq2_normals">
			<param name="LOOP_RATE" value="10.0"/>
			<param name="SEARCH_RADIUS" value="0.5"/>
			<param name="RANDOM_STEP_MAX" value="300"/>
			<param name="THRESHOLD_REF_POINTS" value="50"/>
			<param name="THRESHOLD_ANGLE_FROM_G" value="30.0"/>
			<param name="THRESHOLD_SUM_SQUARE_ERRORS" value="0.001"/>
			<param name="MIN_DISTANCE_BETWEEN_NORMSLS" value="0.1"/>
			<param name="MIN_NUM_GROUPS" value="2"/>
			<param name="FACTOR_1" value="1.0"/>	<!-- num_refpoints -->
			<param name="FACTOR_2" value="0.001"/>	<!-- fitting_error -->
			<param name="FACTOR_3" value="1.0"/>	<!-- ang_from_g_est -->
			<param name="FACTOR_4" value="1.0"/>	<!-- num_groups -->
			<param name="EST_POSEMSG_NAME" value="$(arg posemsg_name)"/>
		</node>

		<!-- <node pkg="beginner_tutorials" name="ekf" type="ekf" output="screen"> -->
		<!-- 	<param name="PUB_POSE_NAME" value="$(arg posemsg_name)"/> -->
		<!-- 	<param name="USE_SLAM" value="true"/> -->
		<!-- 	<param name="USE_WALLS" value="true"/> -->
		<!-- </node> -->

		<!-- <node pkg="beginner_tutorials" name="tf_from_posemsg" type="tf_from_posemsg"> -->
		<!-- 	<param name="SUB_MSG" value="$(arg posemsg_name)"/> -->
		<!-- 	<param name="PARENT_FRAME" value="/odom"/> -->
		<!-- 	<param name="CHILD_FRAME" value="$(arg posemsg_name)"/> -->
		<!-- </node> -->
	</group>

	<!-- <group ns="imu_slam"> -->
	<!-- 	<!&#45;&#45; <node pkg="beginner_tutorials" name="ekf_imu_slam" type="ekf_imu_slam" output="screen"/> &#45;&#45;> -->
	<!-- 	 -->
	<!-- 	<node pkg="beginner_tutorials" name="ekf" type="ekf" output="screen"> -->
	<!-- 		<param name="PUB_POSE_NAME" value="/pose_imu_slam"/> -->
	<!-- 		<param name="USE_SLAM" value="true"/> -->
	<!-- 		<param name="USE_WALLS" value="false"/> -->
	<!-- 	</node> -->
    <!--  -->
	<!-- 	<node pkg="beginner_tutorials" name="tf_from_posemsg" type="tf_from_posemsg"> -->
	<!-- 		<param name="SUB_MSG" value="/pose_imu_slam"/> -->
	<!-- 		<param name="PARENT_FRAME" value="/odom"/> -->
	<!-- 		<param name="CHILD_FRAME" value="/pose_imu_slam"/> -->
	<!-- 	</node> -->
	<!-- </group> -->

	<!-- <node pkg="beginner_tutorials" name="tf_lsdslam" type="tf_lsdslam"/> -->
	
	<node pkg="beginner_tutorials" name="dualekf" type="dualekf" output="screen"/>

	<!-- <node pkg="beginner_tutorials" name="odom2dto3d" type="odom2dto3d"/> -->
	<!-- <node pkg="beginner_tutorials" name="odom2dto3d_" type="odom2dto3d_"/> -->
	<node pkg="beginner_tutorials" name="odom3d" type="odom3d"/>
	<node pkg="beginner_tutorials" name="gyrodometry3d" type="gyrodometry3d"/>
	<node pkg="beginner_tutorials" name="lsdodom" type="lsdodom"/>

	<!-- <node pkg="beginner_tutorials" name="realtimegraph.py" type="realtimegraph.py"/> -->
	<node pkg="beginner_tutorials" name="graph_rpy.py" type="graph_rpy.py"/>
	<!-- <node pkg="beginner_tutorials" name="graph_imu.py" type="graph_imu.py"/> -->

	<!-- <node pkg="rosbag" type="play" name="player" args="&#45;&#45;clock /home/amsl/ozaki/bagfiles/2018&#45;05&#45;08&#45;19&#45;45&#45;05.bag"/> -->
	<!-- <node pkg="rosbag" type="play" name="player" args="&#45;&#45;clock /home/amsl/ozaki/bagfiles/sqlidar_realsense/2018&#45;08&#45;27&#45;11&#45;32&#45;51.bag"/> -->
	<!-- <node pkg="rosbag" type="play" name="player" args="&#45;&#45;clock /home/amsl/ozaki/bagfiles/sqlidar_realsense/2018&#45;09&#45;13&#45;13&#45;13&#45;56.bag"/> -->
	<!-- <node pkg="rosbag" type="play" name="player" args="&#45;&#45;clock /home/amsl/ozaki/bagfiles/sqlidar_realsense/2018&#45;09&#45;14&#45;14&#45;34&#45;50.bag"/> -->
	<!-- <node pkg="rosbag" type="play" name="player" args="&#45;&#45;clock /home/amsl/ozaki/bagfiles/sqlidar_realsense/2018&#45;09&#45;22&#45;12&#45;08&#45;36.bag"/> -->
	
	<!-- <node pkg="rosbag" type="play" name="player" args="&#45;&#45;clock /home/amsl/ozaki/bagfiles/sq2/2018&#45;09&#45;24&#45;14&#45;07&#45;12.bag"/> -->
	<node pkg="rosbag" type="play" name="player" args="--clock /home/amsl/ozaki/bagfiles/sq2/2018-09-24-14-33-49.bag"/>

	<node pkg="rviz" type="rviz" name="rviz" args="-d /home/amsl/.rviz/pose_comparison.rviz"/>
</launch>
